File: backend/app/schemas/due.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import Field

from app.schemas.base import BaseSchema

INTERVAL_PATTERN = "^(yearly|half_yearly|quarterly|monthly|one_time)$"


# --- FeeType ---


class FeeTypeCreate(BaseSchema):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=5000)
    amount: Decimal = Field(ge=0, max_digits=10, decimal_places=2)
    interval: str = Field(default="yearly", pattern=INTERVAL_PATTERN)
    is_active: bool = True


class FeeTypeUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=5000)
    amount: Decimal | None = Field(default=None, ge=0, max_digits=10, decimal_places=2)
    interval: str | None = Field(default=None, pattern=INTERVAL_PATTERN)
    is_active: bool | None = None


class FeeTypeResponse(BaseSchema):
    id: uuid.UUID
    name: str
    description: str | None = None
    amount: Decimal
    interval: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


# --- MemberFee (assignment) ---


class MemberFeeCreate(BaseSchema):
    member_id: uuid.UUID
    fee_type_id: uuid.UUID
    valid_from: date
    valid_to: date | None = None
    note: str | None = Field(default=None, max_length=5000)


class MemberFeeUpdate(BaseSchema):
    valid_from: date | None = None
    valid_to: date | None = None
    note: str | None = Field(default=None, max_length=5000)


class MemberFeeResponse(BaseSchema):
    id: uuid.UUID
    member_id: uuid.UUID
    fee_type_id: uuid.UUID
    valid_from: date
    valid_to: date | None = None
    note: str | None = None
    created_at: datetime
    updated_at: datetime


# --- Due ---


class DueGenerateRequest(BaseSchema):
    year: int = Field(ge=2000, le=2100)


class DuePayRequest(BaseSchema):
    paid_at: date | None = None  # defaults to today in service
    payment_method: str | None = Field(default=None, max_length=50)
    note: str | None = Field(default=None, max_length=5000)


class DueUpdate(BaseSchema):
    note: str | None = Field(default=None, max_length=5000)
    due_date: date | None = None


class DueResponse(BaseSchema):
    id: uuid.UUID
    member_id: uuid.UUID
    member_name: str | None = None
    fee_type_id: uuid.UUID
    fee_name: str
    amount: Decimal
    period_start: date
    period_end: date
    due_date: date
    status: str
    paid_at: date | None = None
    payment_method: str | None = None
    note: str | None = None
    created_at: datetime
    updated_at: datetime


class DueSummaryResponse(BaseSchema):
    open_count: int
    open_amount: Decimal
    paid_count: int
    paid_amount: Decimal

File: apps/web/lib/types/due.ts
/** The billing intervals the backend accepts — mirrors `INTERVAL_PATTERN`. */
export const FEE_INTERVAL_KEYS = [
  "yearly",
  "half_yearly",
  "quarterly",
  "monthly",
  "one_time",
] as const

export type FeeInterval = (typeof FEE_INTERVAL_KEYS)[number]

/** The states a due moves through. */
export const DUE_STATUS_KEYS = ["open", "paid", "cancelled"] as const

/** A fee type as returned by `/api/v1/dues/fee-types`. */
export type FeeType = {
  id: string
  name: string
  description: string | null
  amount: string
  interval: string
  is_active: boolean
}

/** A member↔fee assignment as returned by `/api/v1/dues/assignments`. */
export type MemberFeeAssignment = {
  id: string
  member_id: string
  fee_type_id: string
  valid_from: string
  valid_to: string | null
  note: string | null
}

/** Club totals as returned by `/api/v1/dues/summary`. Amounts are strings. */
export type DuesSummary = {
  open_count: number
  open_amount: string
  paid_count: number
  paid_amount: string
}

/**
 * A due as returned by `/api/v1/dues` and `/api/v1/dues/me`.
 *
 * Amounts are strings: the backend serialises Decimal in JSON mode, and money
 * must not pass through a float on the way to the screen.
 */
export type MyDue = {
  id: string
  member_id: string
  member_name: string | null
  fee_type_id: string
  fee_name: string
  amount: string
  period_start: string
  period_end: string
  due_date: string
  status: string
  paid_at: string | null
  payment_method: string | null
  note: string | null
}
